Stack overflow when using recursive JSON schemas

When `equals` or `hashCode` is called on an underlying raw JSON Schema that is recursive, a `StackOverFlowError` results.

```
Unhandled exception resulting in internal server error response
java.lang.StackOverflowError
at java.base/java.util.Arrays.hashCode(Arrays.java:4685)
at java.base/java.util.Objects.hash(Objects.java:146)
at org.everit.json.schema.Schema.hashCode(Schema.java:224)
at org.everit.json.schema.CombinedSchema.hashCode(CombinedSchema.java:224)
at java.base/java.util.Objects.hashCode(Objects.java:116)
at java.base/java.util.HashMap$Node.hashCode(HashMap.java:297)
at java.base/java.util.AbstractMap.hashCode(AbstractMap.java:527)
at java.base/java.util.Collections$UnmodifiableMap.hashCode(Collections.java:1490)
at java.base/java.util.Arrays.hashCode(Arrays.java:4685)
at java.base/java.util.Objects.hash(Objects.java:146)
at org.everit.json.schema.ObjectSchema.hashCode(ObjectSchema.java:358)
at java.base/java.util.Arrays.hashCode(Arrays.java:4685)
at java.base/java.util.Objects.hash(Objects.java:146)
at org.everit.json.schema.ReferenceSchema.hashCode(ReferenceSchema.java:159)
at java.base/java.util.Arrays.hashCode(Arrays.java:4685)
at java.base/java.util.Objects.hash(Objects.java:146)
at org.everit.json.schema.ArraySchema.hashCode(ArraySchema.java:221)
at java.base/java.util.ArrayList.hashCodeRange(ArrayList.java:626)
at java.base/java.util.ArrayList.hashCode(ArrayList.java:613)
at java.base/java.util.Arrays.hashCode(Arrays.java:4685)
at java.base/java.util.Objects.hash(Objects.java:146)
at org.everit.json.schema.CombinedSchema.hashCode(CombinedSchema.java:224)
...
at java.base/java.util.Objects.hashCode(Objects.java:116)
at java.base/java.util.HashMap$Node.hashCode(HashMap.java:297)
at java.base/java.util.AbstractMap.hashCode(AbstractMap.java:527)
at java.base/java.util.Collections$UnmodifiableMap.hashCode(Collections.java:1490)
at java.base/java.util.Arrays.hashCode(Arrays.java:4685)
at java.base/java.util.Objects.hash(Objects.java:146)
at org.everit.json.schema.ObjectSchema.hashCode(ObjectSchema.java:358)
at java.base/java.util.Arrays.hashCode(Arrays.java:4685)
at java.base/java.util.Objects.hash(Objects.java:146)
at org.everit.json.schema.ReferenceSchema.hashCode(ReferenceSchema.java:159)
at java.base/java.util.Arrays.hashCode(Arrays.java:4685)
at java.base/java.util.Objects.hash(Objects.java:146)
[ERROR] 2020-07-23 20:55:58,898 [qtp1363793720-29037] io.confluent.rest.exceptions.DebuggableExceptionMapper createResponse lsrc-ok52x - Request Failed with exception 
java.lang.StackOverflowError
```

--- a/json-schema-provider/src/main/java/io/confluent/kafka/schemaregistry/json/diff/CombinedSchemaDiff.java
+++ b/json-schema-provider/src/main/java/io/confluent/kafka/schemaregistry/json/diff/CombinedSchemaDiff.java
@@ -15,12 +15,14 @@
 
 package io.confluent.kafka.schemaregistry.json.diff;
 
+import io.confluent.kafka.schemaregistry.json.JsonSchema;
 import io.confluent.kafka.schemaregistry.json.utils.Edge;
 import io.confluent.kafka.schemaregistry.json.utils.MaximumCardinalityMatch;
 import java.util.HashSet;
 import java.util.LinkedHashSet;
 import java.util.List;
 import java.util.Set;
+import java.util.stream.Collectors;
 import org.everit.json.schema.CombinedSchema;
 import org.everit.json.schema.CombinedSchema.ValidationCriterion;
 import org.everit.json.schema.Schema;
@@ -42,8 +44,10 @@ class CombinedSchemaDiff {
     Difference.Type type = compareCriteria(ctx, original.getCriterion(), update.getCriterion());
     if (type != COMBINED_TYPE_CHANGED) {
       // Use sets to collapse duplicate entries
-      final Set<Schema> originalSubschemas = new LinkedHashSet<>(original.getSubschemas());
-      final Set<Schema> updateSubschemas = new LinkedHashSet<>(update.getSubschemas());
+      final Set<JsonSchema> originalSubschemas = original.getSubschemas().stream()
+          .map(JsonSchema::new).collect(Collectors.toCollection(LinkedHashSet::new));
+      final Set<JsonSchema> updateSubschemas = update.getSubschemas().stream()
+          .map(JsonSchema::new).collect(Collectors.toCollection(LinkedHashSet::new));
       int originalSize = originalSubschemas.size();
       int updateSize = updateSubschemas.size();
       if (originalSize < updateSize) {
@@ -61,26 +65,26 @@ class CombinedSchemaDiff {
       }
 
       int index = 0;
-      Set<Edge<Schema, List<Difference>>> compatibleEdges = new HashSet<>();
-      for (Schema originalSubschema : originalSubschemas) {
+      Set<Edge<JsonSchema, List<Difference>>> compatibleEdges = new HashSet<>();
+      for (JsonSchema originalSubschema : originalSubschemas) {
         try (Context.PathScope pathScope = ctx.enterPath(original.getCriterion() + "/" + index)) {
-          for (Schema updateSubschema : updateSubschemas) {
+          for (JsonSchema updateSubschema : updateSubschemas) {
             final Context subctx = ctx.getSubcontext();
-            SchemaDiff.compare(subctx, originalSubschema, updateSubschema);
+            SchemaDiff.compare(subctx, originalSubschema.rawSchema(), updateSubschema.rawSchema());
             if (subctx.isCompatible()) {
               compatibleEdges.add(
-                  new Edge(originalSubschema, updateSubschema, subctx.getDifferences()));
+                  new Edge<>(originalSubschema, updateSubschema, subctx.getDifferences()));
             }
           }
         }
         index++;
       }
 
-      MaximumCardinalityMatch<Schema, List<Difference>> match =
+      MaximumCardinalityMatch<JsonSchema, List<Difference>> match =
           new MaximumCardinalityMatch<>(compatibleEdges, originalSubschemas, updateSubschemas);
-      Set<Edge<Schema, List<Difference>>> matching = match.getMatching();
+      Set<Edge<JsonSchema, List<Difference>>> matching = match.getMatching();
 
-      for (Edge<Schema, List<Difference>> matchingEdge : matching) {
+      for (Edge<JsonSchema, List<Difference>> matchingEdge : matching) {
         ctx.addDifferences(matchingEdge.value());
       }
       if (matching.size() < Math.min(originalSize, updateSize)) {

--- a/json-schema-provider/src/main/java/io/confluent/kafka/schemaregistry/json/diff/CombinedSchemaDiff.java
+++ b/json-schema-provider/src/main/java/io/confluent/kafka/schemaregistry/json/diff/CombinedSchemaDiff.java
@@ -25,7 +25,6 @@ import java.util.Set;
 import java.util.stream.Collectors;
 import org.everit.json.schema.CombinedSchema;
 import org.everit.json.schema.CombinedSchema.ValidationCriterion;
-import org.everit.json.schema.Schema;
 
 import static io.confluent.kafka.schemaregistry.json.diff.Difference.Type.COMBINED_TYPE_CHANGED;
 import static io.confluent.kafka.schemaregistry.json.diff.Difference.Type.COMBINED_TYPE_EXTENDED;

--- a/json-schema-provider/src/test/java/io/confluent/kafka/schemaregistry/json/JsonSchemaTest.java
+++ b/json-schema-provider/src/test/java/io/confluent/kafka/schemaregistry/json/JsonSchemaTest.java
@@ -25,6 +25,8 @@ import com.fasterxml.jackson.databind.node.NumericNode;
 import com.fasterxml.jackson.databind.node.ObjectNode;
 import com.fasterxml.jackson.databind.node.TextNode;
 import io.confluent.kafka.schemaregistry.client.rest.entities.SchemaReference;
+import io.confluent.kafka.schemaregistry.json.diff.Difference;
+import io.confluent.kafka.schemaregistry.json.diff.SchemaDiff;
 import java.util.Collections;
 import java.util.HashMap;
 import java.util.Map;
@@ -286,6 +288,29 @@ public class JsonSchemaTest {
     Object envelope = JsonSchemaUtils.envelope(schema, null);
     JsonSchema schema2 = JsonSchemaUtils.getSchema(envelope);
     assertEquals(schema, schema2);
+  }
+
+  @Test
+  public void testRecursiveSchema() {
+    String schema = "{\n"
+        + "  \"$schema\": \"http://json-schema.org/draft-07/schema#\",\n"
+        + "  \"$id\": \"task.schema.json\",\n"
+        + "  \"title\": \"Task\",\n"
+        + "  \"description\": \"A task\",\n"
+        + "  \"type\": [\"null\", \"object\"],\n"
+        + "  \"properties\": {\n"
+        + "    \"parent\": {\n"
+        + "        \"$ref\": \"task.schema.json\"\n"
+        + "    },    \n"
+        + "    \"title\": {\n"
+        + "        \"description\": \"Task title\",\n"
+        + "        \"type\": \"string\"\n"
+        + "    }\n"
+        + "  }\n"
+        + "}";
+    JsonSchema jsonSchema = new JsonSchema(schema);
+    List<Difference> diff = SchemaDiff.compare(jsonSchema.rawSchema(), jsonSchema.rawSchema());
+    assertEquals(0, diff.size());
   }
 
   private static Map<String, String> getJsonSchemaWithReferences() {

